src/core/ext/transport/chaotic_good/client/chaotic_good_connector.cc

```diff
@@ -95,65 +97,48 @@ absl::StatusOr<ConnectPromiseEndpointResult> ResultFromHandshake(
   if (chaotic_good_ext != nullptr) {
     chaotic_good_ext->EnableStatsCollection(/*is_control_channel=*/true);
     chaotic_good_ext->UseMemoryQuota(ResourceQuota::Default()->memory_quota());
   }
-  handshake_mgr->DoHandshake(
-      OrphanablePtr<grpc_endpoint>(
-          grpc_event_engine_endpoint_create(std::move(*endpoint))),
-      channel_args, deadline, nullptr /* acceptor */,
-      [result_latch = std::move(result_latch),
-       handshake_mgr](absl::StatusOr<HandshakerArgs*> result) {
-        result_latch->Set(ResultFromHandshake(std::move(result)));
-      });
+  return ConnectPromiseEndpointResult{
+      PromiseEndpoint(std::move(endpoint), std::move(args->read_buffer)),
+      args->args};
 }
 
-auto ConnectPromiseEndpoint(EventEngine::ResolvedAddress addr,
-                            const ChannelArgs& channel_args,
-                            Timestamp deadline) {
+Promise<absl::StatusOr<ConnectPromiseEndpointResult>> ConnectPromiseEndpoint(
+    EventEngine::ResolvedAddress addr, ChannelArgs channel_args,
+    Timestamp deadline) {
   auto event_engine = channel_args.GetObjectRef<EventEngine>();
   auto result_latch = std::make_shared<
       InterActivityLatch<absl::StatusOr<ConnectPromiseEndpointResult>>>();
   auto handshake_mgr = MakeRefCounted<HandshakeManager>();
-  auto connect_hdl = event_engine->Connect(
-      [result_latch, channel_args, handshake_mgr,
-       deadline](absl::StatusOr<std::unique_ptr<EventEngine::Endpoint>>
-                     endpoint) mutable {
-        ExecCtx exec_ctx;
-        OnConnect(std::move(endpoint), std::move(handshake_mgr), channel_args,
-                  deadline, std::move(result_latch));
-      },
-      addr, ChannelArgsEndpointConfig(channel_args),
-      ResourceQuota::Default()->memory_quota()->CreateMemoryAllocator(
-          "data_endpoint_connection"),
-      std::chrono::seconds(kTimeoutSecs));
+  absl::StatusOr<std::string> address =
+      grpc_event_engine::experimental::ResolvedAddressToURI(addr);
+  if (!address.ok()) {
+    return Immediate<absl::StatusOr<ConnectPromiseEndpointResult>>(
+        address.status());
+  }
+  channel_args = channel_args.Set(GRPC_ARG_TCP_HANDSHAKER_RESOLVED_ADDRESS,
+                                  address.value());
+  CoreConfiguration::Get().handshaker_registry().AddHandshakers(
+      HandshakerType::HANDSHAKER_CLIENT, channel_args, nullptr,
+      handshake_mgr.get());
+  handshake_mgr->DoHandshake(
+      nullptr, channel_args, deadline, nullptr /* acceptor */,
+      [result_latch, handshake_mgr](absl::StatusOr<HandshakerArgs*> result) {
+        result_latch->Set(ResultFromHandshake(std::move(result)));
+      });
   return OnCancel(
       [result_latch, await = result_latch->Wait()]() { return await(); },
-      [handshake_mgr, connect_hdl, event_engine]() {
+      [handshake_mgr, event_engine]() {
         handshake_mgr->Shutdown(absl::CancelledError());
-        event_engine->CancelConnect(connect_hdl);
```

Comment:
One thing to note here is that the TCP connect handshaker currently uses the iomgr API, which doesn't provide a way to cancel an in-flight connection attempt.  We have an ugly hack in place there where it pretends to cancel it and then waits asynchronously for the connection attempt to complete.

Hopefully, this won't cause any problems, but I wanted to call it out just so you're aware of it.

Reply:
I think it'll be ok here